### ISSUE TITLE
[Android TV] Directional navigation event listener (TVEventHandler) not working

### ISSUE THREAD

**Issue by yuhaoju** [facebook/react-native#19917]
## Environment

Environment:
  OS: macOS High Sierra 10.13.3
  Node: 8.9.4
  Yarn: 1.5.1
  npm: 5.6.0
  Watchman: 4.9.0
  Xcode: Not Found
  Android Studio: 3.0 AI-171.4443003

Packages: (wanted => installed)
  react: 16.3.1 => 16.3.1
  react-native: 0.55.4 => 0.55.4

## Description
I enabled `TVEventHandler` in my Android TV app with focusable elements, but can't get any directional navigation events from it. None of the `eventType` is `left/right/up/down'. It can only emit `focus/blur` events.

In the documentation, it says there is a class `ReactAndroidTVRootViewHelper` to handle focusable elements navigation events, but I can’t find it in RN source code. In this not merged [commit](https://github.com/facebook/react-native/pull/16500/commits/6dc6bb5d8a590d0eae4f9fc34b0bcb4c62927104) I found `ReactAndroidTVViewManager.java` which may be used to handle navigation events. But it's not in the RN current codebase. Maybe there are other components handling that?

## Reproducible Demo
code sample
```JavaScript
<FlatList
  data={[
    [{id: 10}, {id: 11}, {id: 12}, {id: 13}],
    [{id: 20}, {id: 21}, {id: 22}, {id: 23}],
  ]}
  renderItem={({item: rowItems, index: rowIndex}) => {
    return (
      <ScrollView horizontal={true}>
        {
          rowItems.map((id, index) => {
            return (
              <TouchableHighlight
                key={index}
                onPress={(e) => this.handleTitlePress(e)}
                onShowUnderlay={() => this.handleTitleActive(rowIndex, index)}
              >
                <Text>{id}</Text>
              </TouchableHighlight>
            )
          })
        }
      </ScrollView>
    );
  }}
>
</FlatList>
```

**Comment by reidisaki** [comment]
@Weetbix did you ever get that bug 20100 fixed? the focus engine for RN on androidtv/tvos/firetv is lacking.  if I have some  views laid out as X's. and I'm on the 2nd index and I click right, i watn to focus up tot he 3rd index which is one row higher but I can't seem to get the focus to happen like that. on Android it's even worse than tvOS.
                X    X    X
  X   X

I've tried to use setNativeProps() to trigger a specific element to be focused but there is a 2 second delay between the method call and actually setting the focus.  I've asked this on Stackoverflow but no responses. Any help woudl be GREATLY appreciated!  Thank you!

**Comment by Weetbix** [comment]
@reidisaki No not yet, it auto-closed as it wasn't with the latest version of RN. This weekend I will reproduce it in the latest and reopen the issue. AFAICT react native is not involved in the navigation system which all seems to be handled by android - but native android apps with scroll views seem to work perfectly, maybe its down to the difference in the views that are created in RN vs native android.

It sounds like you *want* a kind of behaviour like in 20100? 

According to this comment https://github.com/facebook/react-native/issues/20181#issuecomment-404988268
You might be able to try setting hasTVPreferred focus when you get to the end of the list. 

I will also take another stab at fixing this issue, as it seems easy to add the direction keys but ran into some trouble last time. If the directional keys are working we can write our own navigation code (hacky but at least it will be functional). 

I would keep in mind that there has only been 1 PR for android TV support so far, so it still has some distance to go :)

**Comment by reidisaki** [comment]
@Weetbix thanks! I have used hasTVPreferred focus and that does get focus but its very cumbersome to use and I noticed the TVEventHandler really only works for ios as androids "onHWKeyEvent" only listens to button presses and not dpad presses so if I want to select an item that is adjacent to an item but in a higher row I can't intercept the dpad events to know what the user is pressing. 

I think you are correct in that we would have to write our own navigation code to handle dpad presses.  I also saw your 20100 pr and yes I can replicate that issue and that would fix some of the issues we're having.
